### DISCUSSION TITLE
Svench and Svelte-kit

### DISCUSSION THREAD

**Discussion by yanick** [rixo/svench#33] [Ideas]
originally at https://github.com/rixo/svench/pull/32#issuecomment-868022337

Hey! Thank you very much for this. I'm sorry I have so little time for everything these days... :-/

I've started to look at your PR but got sidetracked by a terrifying bug that turned out to be what looks like incompatibility between HMR and Svelte dev tools... Anyway, I'll throw a few comments and ideas to move it forward.

    And I'm not 100% sure that should go in the docs package vs being in svench/docs

Yeah the docs package is the way to go. I intend to have it deployed at https://svench.dev at some point -- but still having difficulties with pnpm + Vercel... The docs folder in the svench package are more notes to help with documenting later... Actually, the closer to actual usage docs that exists in the project might be examples/rollup, that was the original example, and served a lot as a test bed as I was developing new features.

    could we have a svelte-kit route that looks something like

Yes and no. At its heart, the runtime part of Svench is indeed implemented as a Svelte component. So you can do something like this:

<script>
  import { Svench } from 'svench'
  import routes from './routes.js'
  import ui from './ui.js'
</script>

<Svench {routes} {ui} />

That's kinda how Svench works actually, except the component is created with new Svench(...).

The ui prop contains Svelte components for App (menu, top bar, panels, etc.), and the UI around View. I'd like the UI to be "pluggable" via injection of custom components, eventually.

The routes are normally extracted from your filesystem by routix (a Node program that watch FS and regenerates a tree structure from it). You could theoretically generate the routes object by other means, or maybe run Routix separately. Check the .svench/src/routes.js that is generated in your project to get an idea (and maybe the svench.js file in there too, that is the entry point of your Svench app).

There would surely be some adaptations needed to support really being able to do this... Now the biggest problem if you want to put that in a Kit app is that the two will fight over client side routing (Svench's own router is a slightly tweaked version of Navaid).

What would you expect from having Svench's router in a Kit route?

I'm interesting in pursuing this discussion, or others, but I don't think we should do it here. This PR will get closed soon, I hope. I have enabled the Discussion tab in this repo, so that might be a place. Or maybe a real issue, if there's a real use case / feature to consider.

**Comment by yanick** [comment]
> Hey! Thank you very much for this. I'm sorry I have so little time for everything these days... :-/

Don't worry. I maintain a few OS projects of my own, so I know painfully well how the hours of the day doesn't match the time required to do all the things.

> I've started to look at your PR but got sidetracked by a terrifying bug that turned out to be what looks like incompatibility between HMR and Svelte dev tools...

:scream: 


> Yeah the docs package is the way to go. [..] The docs folder in the `svench` package are more notes to help with documenting later... [..] the closer to actual usage docs that exists in the project might be `examples/rollup`

Gotcha x 3!
 
> > could we have a svelte-kit route that looks something like
> 
> Yes and no. At its heart, the runtime part of Svench is indeed implemented as a Svelte component. So you can do something like this: [..]

Thanks a lot for that explanation. It really helps to get a grasp of the pieces involved, and how they play together.

> What would you expect from having Svench's router in a Kit route?

I guess the first thing I should say is that I'm mostly leisurely toying with things. There is not a grand masterplan (yet), just an interest to pop the hood of that gorgeous project open and see how it ticks underneath.

This being said, the two ideas that made want to try Svench-within-kit are:

1. having Svench-within-kit should take care of having Svench configurations (directory aliases, vite options, etc) match svelte-kit's as the former would vanish and be superceded by the latter. 
2. that would also make the generation of a static version of the Svench pages via `svelte-kit build`.
 
that will probably come with the price of having to manually declare the svench components in the `routes` prop of `Svench` (so basically do manually what routix is currently doing for us). Which might not be too horrendous for small projects -- it's not like there are bazillions of svench files to import, and most of the hot-reloading would be covered by svelte-kit's own system.

But yeah, right now it's all half-baked ideas and some strong urge to experiment. If I ever come to something that actually makes sense and feels like it could be useful, be sure I'll let you know. ^.^

**Comment by rixo** [comment]
> having Svench-within-kit should take care of having Svench configurations (directory aliases, vite options, etc) match svelte-kit's as the former would vanish and be superceded by the latter.

I don't really believe in Kit support via Svench as Kit app... The first prototype of Svench was developed over Routify, which is arguably more "exposed" than Kit and also much simpler (at least at the time, and I'm only referring to the router parts). Even then, Routify is optimized for different goals than Svench. Namely Routify focus on producing a performant prod app and also provide dev abstraction aimed at supporting this end goal. The range of use cases covered by Routify is pretty wide, and offering tools to address them all explains a good deal of Routify's complexity (as in the "complexity / feature" trade off, I'm not saying Routify is needlessly complicated). Same basically apply to Kit. On the other hand, Svench's primary goal is comparatively very narrow and, as a matter of fact, somewhat accessory for an app building framework: Svench is all about shortening the path between a change in your editor and preview in the browser during dev.

For those reasons, as Svench prototyping progressed, Routify's features were getting more and more "in the way" of Svench's primary goal. In particular, I met issues with the routes generator, that was hard to customize because it needed to take into account all the other features offered by Routify (and often not useful for Svench). That's why I developed a lightweight version of it (Routix), with less features, but the primary target from the ground up to be able to do incremental generation. Also, Routify was getting in the way of HMR. Unfortunately, HMR remains a very tricky act, and mundane things can bring it down or make it much less useful. Svench is willing to sacrifice features, or make its internal code more complicated, toward helping HMR work well, which would make much less sense for Routify or Kit...

That's why, in the end, I believe the best path toward Kit support is for Svench to know about Kit peculiarities and offer specific adapters and helpers to support them. When I'll get back to it, I'll try to consider your idea and see whether Kit could somehow be substituted to Vite, and if we could get some simplification from it. But I doubt it because, for one thing Kit, as a dev server, is a purposefully limited subset of Vite -- actually Kit actively aims at abstracting Vite as an implementation detail, and, secondly, even if it was somehow possible, Kit would have to be severely tweaked for that to work and, in the end we would have no guarantee that this tweaked setup would still work for or represent the user's app environment.

The use cases I'd expect from using Svench components directly in a Kit app would be more something around having a (maybe heavily customized) Svench style guide or so in some part of a Kit app. But in this case, I think the strategy would probably go toward isolating them from each other to the maximum, to limit friction / conflict surface.

> that would also make the generation of a static version of the Svench pages via svelte-kit build.

_That_ is already provided by Vite (and wrapped by Svench):

```bash
pnpx svench build
```

That's what is already used to push the docs to Vercel: https://svench-docs-pub6uad4j-rixodev.vercel.app/?fs=0&c&o=0&p&cv=%40none&bg=%23fff (unfortunately the autodeploy part doesn't work smoothly currently because of Vercel not liking pnpm's workspace dep in the docs package :-/)

The end result of the build is probably much less polished than what you get with Routify of Vite, in matter of code splitting, pretty URLs, pre-fetching and all, which is the result of the opposite trade offs made by Svench vs app frameworks. But you still get to build & publish your workbench, if you want to do so (and still, Svench has its own goal of code splitting to the max for fastest (re)loading times during dev, so you do get that in the prod build). You also don't get all the fancy adapters that Kit offers but, here again, that doesn't seem very useful for Svench. In most cases, Svench only ever wants to export to static. It is not expected to ever have complex data fetching or server side problematic (that would go contrary to the goal of isolating components), that are the reasons why Kit needs adapters.

**Reply by yanick** [reply]
> I don't really believe in Kit support via Svench as Kit app... [..]

Thanks for that explanation, very much appreciated.

And yeah, I don't have a good enough grasp of all the intricacies of Svench (and Svelte-kit, and Vite, for what matters) to be anything more than the usual bumbling fool that goes "hey, wouldn't it be easier to [insert something that, no, wouldn't be easier]". So I totally bow down to your opinion on the matter. That gives me a lot of ponder upon, and... well, it won't stop me from experimenting, but that's what bumbling fools are wont to do. ^.^


> > that would also make the generation of a static version of the Svench pages via svelte-kit build.
> 
> _That_ is already provided by Vite (and wrapped by Svench):
> 
> ```shell
> pnpx svench build
> ```

Oooh.... I have to try that (and add it to the documentation when successful!)

> That's what is already used to push the docs to Vercel: https://svench-docs-pub6uad4j-rixodev.vercel.app/?fs=0&c&o=0&p&cv=%40none&bg=%23fff (unfortunately the autodeploy part doesn't work smoothly currently because of Vercel not liking pnpm's workspace dep in the docs package :-/)

For what it's worth, I'm a big fan of pnpm. Mostly because I have relatively small hard disks and a ridiculous amount of projects lying around. The npm way of copying all dependencies for each and every project always make me twitch a little bit. ^.^


> The end result of the build is probably much less polished than what you get with Routify of Vite, in matter of code splitting, pretty URLs, pre-fetching and all,  

I would think that's fine. I mean, having everything optimized is nice and all, but Svench and its output are fundamentally development tools that are mostly going to live behind the curtain. So if that output is going to be slightly bigger and clunkier than the app generated by Routify/Svelte-kit, meh, it's fine -- better to have a chunkier toy than nothing at all!

**Comment by dipscom** [comment]
Just would like to say, I'm understanding very little of the technical side of it. But I am loving reading the exchange. Please, carry on. If you two carry on for long enough and I repeately read the same exchange several times over, I might, one day, understand something.
